MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php

```diff
@@ -0,0 +1,163 @@
+ */
+class GetTree extends Action
+{
+    private const HTTP_OK = 200;
+    private const HTTP_INTERNAL_ERROR = 500;
+
+    /**
+     * @see _isAllowed()
+     */
+    public const ADMIN_RESOURCE = 'Magento_MediaGalleryUi::media_gallery';
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var Filesystem
+     */
+    private $filesystem;
+
+    /**
+     * @var string
+     */
+    private $path;
+
+    /**
+     * @var array
+     */
+    private $responseContent;
+
+    /**
+     * Constructor
+     *
+     * @param Action\Context $context
+     * @param LoggerInterface $logger
+     * @param Filesystem $filesystem
+     * @param string $path
+     */
+    public function __construct(
+        Action\Context $context,
+        LoggerInterface $logger,
+        Filesystem $filesystem,
+        string $path
+    ) {
+        parent::__construct($context);
+        $this->logger = $logger;
+        $this->filesystem = $filesystem;
+        $this->path = $path;
+        $this->responseContent = [];
+    }
+    /**
+     * @inheritdoc
+     */
+    public function execute()
+    {
+        try {
+            $directories = [];
+            $directoryInstance = $this->filesystem->getDirectoryRead($this->path);
+            if ($directoryInstance->isDirectory()) {
+                foreach ($directoryInstance->readRecursively() as $index => $path) {
+                    if ($directoryInstance->isDirectory($path)) {
+                        $pathArray = explode('/', $path);
+                        $directories[] =
+                            [
+                                'data' => count($pathArray) > 0 ? end($pathArray) : $path,
+                                'attr' => ['id' => $index],
+                                'metadata' => [
+                                    'path' => $path
+                                ],
+                                'path_array' => $pathArray
+                            ];
+                    }
+                }
+            }
+            $this->responseContent = $this->buildFolderTree($directories);
+
+            $responseCode = self::HTTP_OK;
+        } catch (\Exception $exception) {
+            $this->logger->critical($exception);
+            $responseCode = self::HTTP_INTERNAL_ERROR;
+            $this->responseContent = [
+                'success' => false,
+                'message' => __('Retrieving directories list failed.'),
+            ];
+        }
+
+        /** @var Json $resultJson */
+        $resultJson = $this->resultFactory->create(ResultFactory::TYPE_JSON);
+        $resultJson->setHttpResponseCode($responseCode);
+        $resultJson->setData($this->responseContent);
+
+        return $resultJson;
+    }
+
+    /**
+     * Build folder tree structure by provided directories path
+     *
+     * @param array $directories
+     * @return array
+     */
+    private function buildFolderTree(array $directories, bool $skipRoot = true): array
```

Comment:
Is it possible to perform the final folder tree creation in the directory iterator loop? (for better performance)